### PR TITLE
[Merged by Bors] - refactor(Tactic/Linarith): introduce `UsableInSimplexAlgorithm` class to allow the use of sparse matrix types in the oracle

### PR DESCRIPTION
**Pull request by vasnesterov**
This is the first part of #12819, which I decided to split into two parts. The initial goal is to enable the use of sparse matrices in the `linarith`'s oracle based on the Simplex Algorithm, which speeds up `linarith`, and especially `nlinarith`, in many cases.

* Introduce the `UsableInSimplexAlgorithm` class, which specifies the operations required from a matrix type for use in Gauss Elimination and the Simplex Algorithm within the corresponding oracle for the `linarith` tactic. Rewrite these algorithms to use this class, allowing any matrix type to be used.
* Make `Matrix` an instance of this class.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by vasnesterov** [comment]
!bench

**Review by robertylewis — COMMENTED** [review]
This looks like a totally reasonable generalization!

**Comment by robertylewis** [comment]
(I should add: I completely missed your PR a few month ago implementing the simplex oracle and changing linarith over. This is fantastic work and really appreciated!)

File: Mathlib/Tactic/Linarith/SimplexAlgorithm/Datatypes.lean [hunk 1]
```diff
@@ -24,24 +48,32 @@ store matrix as its entries than as function between `Fin`-s.
...
+  swapRows mat i j := ⟨mat.data.swap! i j⟩
+  subtractRow mat i j coef :=
+    let newData : Array (Array Rat) := mat.data.modify j fun row =>
+      row.zipWith mat.data[i]! fun x y => x - coef * y
+    ⟨newData⟩
+  divideRow mat i coef := ⟨mat.data.modify i (·.map (· / coef))⟩
 
 /--
-`Table` is a structure Simplex Algorithm operates on. The `i`-th row of `mat` expresses the
+`Tableu` is a structure the Simplex Algorithm operates on. The `i`-th row of `mat` expresses the
```
Comment (robertylewis) [inline Mathlib/Tactic/Linarith/SimplexAlgorithm/Datatypes.lean:68]:
```suggestion
`Tableau` is a structure the Simplex Algorithm operates on. The `i`-th row of `mat` expresses the
```
and below? (c.f. https://en.wikipedia.org/wiki/Simplex_algorithm#Simplex_tableau)

(Sorry, for some reason this didn't get submitted yesterday. This is my only request, then I'll merge.)

Comment (vasnesterov) [reply Mathlib/Tactic/Linarith/SimplexAlgorithm/Datatypes.lean]: Sorry for misspelling this twice :)

**Comment by robertylewis** [comment]
Thanks!

bors merge

**Pull request closed without merging**